src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,259 @@
+ * temporarily process more messages than it receives. Similarly, cowns sending
+ * to muted receivers will also be muted so that cowns don't experience runaway
+ * queue growth while they are muted. Muted cowns are generally not rescheduled
+ * until the receiver that resulted in their muting (the "mutor") is no longer
+ * muted or overloaded.
+ *
+ * ## Cown Load
+ *
+ * The backpressure system uses "load" as an approximation for message queue
+ * depth. Load is measured using a token message that occasionally falls out of
+ * its message queue. The "current load" on a cown is equal to the number of
+ * messages processed by the cown since the last token message fell out. Once
+ * the next token falls out of its queue, the current load is stored into a ring
+ * buffer and the current load is reset. The total load for the cown is
+ * calculated as the sum of the current load and previous loads stored in the
+ * ring buffer. Simply measuring the batch size of messages processed by the
+ * cown does not account for additional participants of a message cutting their
+ * batch short once they receive the mutli-message.
+ *
+ * ## Mute Map Scan
+ *
+ * Once a scheduler thread completes a message action that would result in the
+ * muted multimessage cowns, the cowns are acquired by the scheduler thread and
+ * placed in a "mute map" on the scheduler thread. The mute map is a mapping
+ * from mutor => mute set, where "mute set" referrs to the set of cowns muted by
+ * a mutor. Each scheduler thread scans the mutors of its mute map on each
+ * iteration of the run loop. If any mutor is not muted and is not overloaded,
+ * the corresponding mute set is unmuted and the entry is removed from the mute
+ * map. A mutor may exist as the key in mutiple mute maps, but a muted cown may
+ * only be tracked in a single mute set.
+ *
+ * ## Message Scan
+ *
+ * All messages from one set of sending cowns to a set of receiving cowns is
+ * scanned to determine if the senders should be muted. Such a message will
+ * result in muted senders if all of the following are true:
+ *   1. The receiver set does not contain any of the senders
+ *   2. No sender is overloaded
+ *   3. One of the receivers is either overloaded or muted.
+ *
+ * A message is given "priority" if either condition 1 or 2 are false. A
+ * priority message will result in any muted receivers being unmuted so that the
+ * senders may make progress sooner.
+ *
+ * The backpressure bits of any cown may only be modified by the scheduler
+ * thread running the cown or holding the muted cown in its mute map. If a
+ * priority message requires a recipient cown to be unmuted, then the scheduler
+ * thread running the message will search for the cown in its mute map. If the
+ * cown does not exist, then it will be sent to other scheduler threads until it
+ * is found and subsequently unmuted.
+ *
+ * ## Cown Backpressure States
+ *
+ * These states are not explicitly encoded, but they are useful to consider.
```

Comment:
I don't understand this section.

Reply:
Yeah, this is a holdover from a previous design where these states were more explicit. I think I'll remove them since they may just lead to confusion.